src/QMCDrivers/QMCDriverNew.cpp

```diff
@@ -485,37 +468,60 @@ std::ostream& operator<<(std::ostream& o_stream, const QMCDriverNew& qmcd)
 
 void QMCDriverNew::defaultSetNonLocalMoveHandler(QMCHamiltonian& ham) {}
 
-QMCDriverNew::AdjustedWalkerCounts QMCDriverNew::adjustGlobalWalkerCount(Communicate* comm,
-                                                                         IndexType desired_count,
+QMCDriverNew::AdjustedWalkerCounts QMCDriverNew::adjustGlobalWalkerCount(int num_ranks,
+                                                                         int rank_id,
+                                                                         IndexType required_total,
                                                                          IndexType walkers_per_rank,
                                                                          RealType reserve_walkers,
                                                                          int num_crowds)
 {
-  int ranks = myComm->size();
-  AdjustedWalkerCounts awc{0, 0, 0, 0};
-  awc.global_walkers   = desired_count;
-  awc.walkers_per_rank = walkers_per_rank;
-  awc.reserve_walkers  = reserve_walkers;
-  if (awc.global_walkers != 0)
+  AdjustedWalkerCounts awc{required_total, {}, {}, reserve_walkers};
+
+  if (required_total != 0)
   {
-    if (awc.global_walkers % ranks)
+    awc.walkers_per_rank = fairDivide(required_total, num_ranks);
+    if (walkers_per_rank != 0)
     {
-      awc.walkers_per_rank = awc.global_walkers / ranks + 1;
-      awc.global_walkers   = awc.walkers_per_rank * ranks;
-      app_warning() << "TargetWalkers not divisible by number of ranks, TargetWalker increased to "
-                    << awc.global_walkers << '\n';
+      if (std::any_of(awc.walkers_per_rank.begin(), awc.walkers_per_rank.end(),
+                      [walkers_per_rank](IndexType num_walkers) { return num_walkers != walkers_per_rank; }))
+      {
+        std::ostringstream error;
+        error << "Running on " << num_ranks << " and the request of " << required_total << " walkers and "
+              << walkers_per_rank << " cannot be satisfied.";
+        throw std::runtime_error(error.str());
+      }
     }
-    else
-      awc.walkers_per_rank = awc.global_walkers / ranks;
+  }
+  else if (walkers_per_rank != 0)
+  {
+    awc.walkers_per_rank = std::vector<IndexType>(num_ranks, walkers_per_rank);
   }
 
-  awc = this->calcDefaultLocalWalkers(awc);
-
-  if (awc.walkers_per_rank * ranks != awc.global_walkers)
+  // if we found a way to determine walkers per rank above
+  if (awc.walkers_per_rank.size() > 0)
   {
-    awc.global_walkers = awc.walkers_per_rank * ranks;
-    app_warning() << "Walkers per rank number of crowds, TargetWalker increased to " << awc.global_walkers << '\n';
+    if (num_crowds == 0)
+      num_crowds = std::min(static_cast<unsigned int>(awc.walkers_per_rank[rank_id]), Concurrency::maxThreads<>());
```

Comment:
I prefer num_crowds not depend on walkers_per_rank[rank_id] but only input value or  Concurrency::maxThreads<>(). https://github.com/ye-luo/qmcpack/wiki/Population-size-input-handling#step-1-reduce-the-relation-map-by-first-handling-num_crowds
I may start the run with 6 walkers_per_rank but 7 crowds.

Could you follow step 1,2,3 order. The order of (a-d) in step 2 doesn't really matter.

Reply:
This doesn't prevent 6  walkers_per_rank and 7 crowds if you specify the num_crowds only if you want to control number of crowds using OMP_NUM_THREADS.  It does prevents the default case on the power 9 development node I use where not setting OMP_MAX_THREADS would result 6 walkers per rank and 176 crowds or with std::thread where the default max_threads is on order of 10k. Since I've done this many times I imagine users will run into this more than they want more crowds than the number of walkers per rank.

Yet it's an extra behavior...  I'll think about this as I take care of other changes.

Reply:
I'd prefer to leave this in unless it really prevents you from doing what you want to with the number of crowds. It prevents a node overloading loading behavior when every hardware thread is part of the OMP_NUM_THREADS default.

Reply:
The current logic is still over-complicated. As you pointed out, I cannot run walkers_per_rank = 6 and OMP_MAX_THREADS=7 with 7 crowds and allow the population to increase from 6 to 7 without a performance cliff which 6 crowds have.

Reply:
I think you should specify num_crowds in that case. I'm not sold on openmp as the one true way, other threading systems do not use environment variables to configure the concurrency environment. Certainly our documentation should not privlege OMP_MAX_THREADS over num_crowds. But I will remove the complexity.

Reply:
Done

Reply:
I meant Concurrency::maxThreads instead of exactly OMP_MAX_THREADS. For any change of the scheme. It is better to figure it out on paper like the wiki and then implemented it exactly. I will move the wiki to the manual since now it has been fully implemented in the batched driver.